File: day/6/c++/main.cpp
#include <fstream>
#include <iostream>
#include <set>

size_t getMarkerLocation(const std::string &stream, size_t block_size) {
    auto packet = std::string(block_size, '0');

    for (size_t i = 0; i < stream.size(); ++i) {
        packet.erase(packet.begin());
        packet += stream[i];

        if (std::set<char>(packet.begin(), packet.end()).size() == block_size)
            return i + 1;
    }
}

int main() {
    auto file = std::ifstream("../input.txt");
    std::string line;
    std::getline(file, line);

    std::cout
        << "Part 1: "
        << getMarkerLocation(line, 4) << std::endl;

    std::cout
        << "Part 2: "
        << getMarkerLocation(line, 14) << std::endl;
}

File: day/6/python/main.py
def get_marker_location(stream, block_size):
    packet = "0" * block_size
    for i, char in enumerate(stream):
        packet = packet[1:] + char
        if len(set(packet)) == block_size:
            return i + 1

with open('../input.txt') as file:
    file = file.read()

    print(f"Part 1: {get_marker_location(stream=file, block_size=4)}")
    print(f"Part 2: {get_marker_location(stream=file, block_size=14)}")
